`prefer-const` false positive

**What version of ESLint are you using?**
2.4.0

**What parser (default, Babel-ESLint, etc.) are you using?**
default

**Please show your full configuration:**
```json
{
  "parserOptions": {
    "ecmaVersion": 6,
    "sourceType": "module",
    "ecmaFeatures": {
      "jsx": true
    }
  },
  "rules": {
    "prefer-const": 2
  }
}
```

**What did you do? Please include the actual source code causing the issue.**
```jsx
import React from 'react';
import { browser } from '../../util/browserUtil';

let DevTools;

if (!browser.ie8()) DevTools = require('../devTools/devTools');

export default ({ children }) => (
  <div>
    <div>
      {children}
    </div>
    {DevTools && <DevTools />}
  </div>
);
```

**What did you expect to happen?**
No warning, as `DevTools` is used outside of the `if`-statement

**What actually happened? Please include the actual, raw output from ESLint.**
```
/home/testyo/repos/meeeeeeep/file.js
  6:5  error  'DevTools' is never reassigned, use 'const' instead  prefer-const
```

I'm aware of many potentially related issues (#5251, #5283, and #5284), but I don't think this one is covered

--- a/docs/rules/prefer-const.md
+++ b/docs/rules/prefer-const.md
@@ -26,6 +26,11 @@ for (let i in [1,2,3]) {
 for (let a of [1,2,3]) {
     console.log(a);
 }
+
+// the initializer is separated.
+let a;
+a = 0;
+console.log(a);
 ```
 
 The following patterns are not considered problems:
@@ -51,6 +56,13 @@ for (const a of [1,2,3]) {
 for (let i = 0, end = 10; i < end; ++i) {
     console.log(a);
 }
+
+// the initializer is located at another block.
+let a;
+if (true) {
+    a = 0;
+}
+console.log(a);
 
 // suggest to use `no-var` rule.
 var b = 3;

--- a/lib/rules/prefer-const.js
+++ b/lib/rules/prefer-const.js
@@ -11,6 +11,7 @@
 //------------------------------------------------------------------------------
 
 var PATTERN_TYPE = /^(?:.+?Pattern|RestElement|Property)$/;
+var DECLARATION_HOST_TYPE = /^(?:Program|BlockStatement|SwitchCase)$/;
 
 /**
  * Adds multiple items to the tail of an array.
@@ -47,7 +48,8 @@ function canBecomeVariableDeclaration(identifier) {
         node.type === "VariableDeclarator" ||
         (
             node.type === "AssignmentExpression" &&
-            node.parent.type === "ExpressionStatement"
+            node.parent.type === "ExpressionStatement" &&
+            DECLARATION_HOST_TYPE.test(node.parent.parent.type)
         )
     );
 }

--- a/tests/lib/rules/prefer-const.js
+++ b/tests/lib/rules/prefer-const.js
@@ -63,6 +63,7 @@ ruleTester.run("prefer-const", rule, {
             parserOptions: { ecmaVersion: 6 }
         },
         { code: "/*exported a*/ let a; function init() { a = foo(); }", parserOptions: { ecmaVersion: 6 } },
+        { code: "let a; if (true) a = 0; foo(a);", parserOptions: { ecmaVersion: 6 } },
 
         // The assignment is located in a different scope.
         // Those are warned by prefer-smaller-scope.
@@ -144,6 +145,11 @@ ruleTester.run("prefer-const", rule, {
 
         {
             code: "let x; x = 0;",
+            parserOptions: { ecmaVersion: 6 },
+            errors: [{ message: "'x' is never reassigned, use 'const' instead.", type: "Identifier"}]
+        },
+        {
+            code: "switch (a) { case 0: let x; x = 0; }",
             parserOptions: { ecmaVersion: 6 },
             errors: [{ message: "'x' is never reassigned, use 'const' instead.", type: "Identifier"}]
         },

